rfcs/convergence/react-focus-management.md

```diff
@@ -0,0 +1,169 @@
+
+Nesting focusables is a problem that ability-helpers explicitly tries to solve, bringing focus management to a lower level than the rendering framework tries to accomplish this consistently independent of rendering cycles.
+
+### Initial API
+
+#### FocusManagementProvider
+
+Context provider that hosts the [ability-helpers](https://github.com/microsoft/ability-helpers) instance and its API objects.
+
+The below table defines what is included in the context value
+
+> Note that this table will increase as more ability-helpers features are included in Fluent
+
+| Name      | Value        | Description                           |
+| --------- | ------------ | ------------------------------------- |
+| focusable | FocusableAPI | API for traversing focusable elements |
+
+The below table defines the props used by the provider component. All context values can be used as props in the provider component but are optional.
+
+| Name                   | Value              | Description                                                                                                           |
+| ---------------------- | ------------------ | --------------------------------------------------------------------------------------------------------------------- |
+| dir                    | 'ltr' \| 'rtl'     | Direction indicator, some keyboard behaviour will need to change based on this value                                  |
+| Deloser                | DeloserAPI         | Tools to help automatically restore focus when it gets lost without manually calling `.focus()` from application code |
+| internal\_\_ahInstance | AbilityHelpersCore | The ability helpers instance, not expected to be be used directly                                                     |
+
+#### Hooks
+
+This spec proposes an API based on React hooks pattern. [ability-helpers](https://github.com/microsoft/ability-helpers) works mainly on adding `data-*` attributes to DOM elements to apply configurable behaviours to the 'groups' of focusables.
+
+However some internal behaviours such as `modalizer` or `deloser` require DOM elements during the configuration phase, which will be most easily configured in React using `refs`.
+
+Therefore it must be noted that **although some hooks might not require ref interaction and can simply return attributes to be spread, in an effort to maintain consistency refs are used, which fits the Fluent component design pattern which enforces ref forwarding**
+
+```typescript
+// Basic up/down keyboard navigation
+const ref = useKeyboardNavigationGroup();
+
+<div ref={ref}>{props.children}</div>;
+```
+
+```typescript
+const ref = React.useReact<HTMLElement>();
+const {
+  findFirst,
+  findLast,
+  findAll,
+} = useFocusable(ref); // All operations are with respect to the given DOM container
```

Comment:
We should avoid APIs that are accepting refs:

```tsx
function App() {
  const ref = React.useRef();
  const [el, setEl] = React.useState();

  // 🟡 will work, but how we will know that `ref` changed?
  // https://github.com/theKashey/use-callback-ref tries to solve this issue, BTW
  useFocusable(ref);

  // 🔴 `el` will become something only after a render cycle, should we re-render? or do not support this case?
  useFocusable(el);

  return (
    <>
      <div ref={ref} /> /* both will work on React components */
      <div ref={setEl} />
    </>
  );
}
```

Reply:
Ah sorry the intention was to accept a DOM element not a ref

Reply:
In this case, an issue related to DOM updates is still there:

```tsx
<MenuItem as={smth ? 'span' : 'div'} /> // 💣 `as` can be a different DOM node
```

In this case, DOM element might change and as we don't control user's input, we will not know about this change. This might not be an issue for this hook, but it's about a pattern in general.

Reply:
updated to use `ref.current` so a DOM element is accepted

Reply:
no more refs for now